src/clustering_quality.jl

```diff
@@ -0,0 +1,304 @@
+    n == k && throw(ArgumentError("Quality index not defined for the degenerated clustering where each data point is its own cluster."))
+    all(>=(0), weights) || throw(ArgumentError("All weights must be larger or equal 0."))
+    1 < fuzziness || throw(ArgumentError("Fuzziness must be greater than 1 ($fuzziness given)"))
+
+    if quality_index == :calinski_harabasz
+        _cluquality_calinski_harabasz(data, centers, weights, fuzziness, metric)
+    elseif quality_index == :xie_beni
+        _cluquality_xie_beni(data, centers, weights, fuzziness, metric)
+    else
+        throw(ArgumentError("Quality index $quality_index not supported."))
+    end
+end
+
+clustering_quality(data::AbstractMatrix{<:Real}, R::FuzzyCMeansResult; quality_index::Symbol, fuzziness::Real, metric::SemiMetric=SqEuclidean()) =
+    clustering_quality(data, R.centers, R.weights; quality_index = quality_index, fuzziness = fuzziness, metric = metric)
+
+
+# clustering indices with cluster centres not known interface
+
+function clustering_quality( 
+        assignments::AbstractVector{<:Integer}, # n vector
+        dist::AbstractMatrix{<:Real};           # n×n matrix
+        quality_index::Symbol 
+    )
+    n, m = size(dist)
+    na = length(assignments)
+    n == m || throw(ArgumentError("Distance matrix must be square."))
+    n == na || throw(DimensionMismatch("Inconsistent array dimensions for distance matrix and assignments."))
+
+    if quality_index == :silhouettes
+        mean(silhouettes(assignments, dist))
+    elseif quality_index == :dunn
+        _cluquality_dunn(assignments, dist)
+    else
+        error(ArgumentError("Quality index $quality_index not available."))
+    end
+end
+
+
+clustering_quality(data::AbstractMatrix{<:Real}, assignments::AbstractVector{<:Integer}; quality_index::Symbol, metric::SemiMetric=SqEuclidean()) = 
+    clustering_quality(assignments, pairwise(metric, data, dims=2); quality_index = quality_index)
+
+clustering_quality(data::AbstractMatrix{<:Real}, R::ClusteringResult;  quality_index::Symbol, metric::SemiMetric=SqEuclidean()) =
+    clustering_quality(R.assignments, pairwise(metric, data, dims=2); quality_index = quality_index)
+
+clustering_quality(R::ClusteringResult, dist::AbstractMatrix{<:Real}; quality_index::Symbol) = 
+    clustering_quality(R.assignments, dist; quality_index = quality_index)
+
+
+# utility functions
+
+function _gather_samples(assignments, k) # cluster_samples[j]: indices of points in cluster j
+    cluster_samples = [Int[] for _ in  1:k]
+    for (i, a) in zip(eachindex(assignments), assignments)
+        push!(cluster_samples[a], i)
+    end
+    return cluster_samples
+end
+
+
+# shared between calinski_harabasz and xie_beni (hard version)
+_inner_inertia(metric::SemiMetric, data::AbstractMatrix, centers::AbstractMatrix,
+               assignments::AbstractVector{<:Integer}) =
```

Comment:
```suggestion
_inner_inertia(metric::SemiMetric, data::AbstractMatrix, centers::AbstractMatrix,
               assignments::AbstractVector{<:Integer}, fuzziness::Nothing) =
```

Reply:
You can calculate inner_inertia without using `_gather_samples` as you suggest, but `colwise` can be more efficient than calling chosen distance within the loop. This makes possible gains unclear and distance-dependent.

Reply:
`pairwise()` is for sure more efficient when we are calculating pairwise distances (for certain metrics) between N and M points.
But in this case it's a distance between N points and 1 point (in `_gather_samples()` implementation too).
Is it also getting significant performance boost?

Reply:
For what I understand, possible gains are entirely metric-dependent. For the most popular one, in
```julia
using BenchmarkTools

X = randn(1000,1000)
C = randn(1000)

@benchmark sum(colwise(Euclidean(),X,C))
@benchmark sum(Euclidean()(x,C) for x in eachcol(X))
```
on my machine I get `colwise` being faster by 60% and taking 10x less memory.

Reply:
Thanks for checking. Could you please add your benchmark results here?
I guess the performance gain is due to the use of BLAS matrix multiplication for Euclidean/SqEuclidean (so for the other metrics the gain should be lower).
The memory inefficiency is a bit unfortunate, I'm not sure it's because of `eachcol(X)` vs `view(X, :, i)`, but due to internal arrays allocation in the generic `metric()` code.
Anyway, thanks again for checking -- you can revert it to the old version (otherwise I will do it), just keep the signature of `inner_inertia()` the same and call `gather_samples()` inside it.

Reply:
Sure, `colwise` is

```julia
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  132.700 μs …  4.223 ms  ┊ GC (min … max): 0.00% … 96.37%
 Time  (median):     137.900 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   144.589 μs ± 47.773 μs  ┊ GC (mean ± σ):  0.28% ±  0.96%

  ▃▇█▆▅▅▃▃▂▂▁▁▁    ▁    ▁                                      ▂
  ██████████████████████████▇▇▆▆▆▇▆▆▆▆▆▅▃▅▅▄▁▄▄▁▄▅▃▁▄▄▃▁▃▄▄▁▅▅ █
  133 μs        Histogram: log(frequency) by time       256 μs <

 Memory estimate: 7.95 KiB, allocs estimate: 2.
```

`Euclidean()(x,C)` is
```julia
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  197.400 μs …   3.395 ms  ┊ GC (min … max): 0.00% … 87.95%
 Time  (median):     203.200 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   211.891 μs ± 122.278 μs  ┊ GC (mean ± σ):  2.17% ±  3.57%

     ▃█▇▇▁         
  ▁▂▅█████▆▄▃▂▂▂▂▂▂▂▂▂▂▂▁▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  197 μs           Histogram: frequency by time          254 μs <

 Memory estimate: 78.17 KiB, allocs estimate: 3001.
```

I've checked `view(X,:,i)` quickly, it is even worse.

Reply:
> I've checked view(X,:,i) quickly, it is even worse.

Worse in terms of memory or performance?

Reply:
On the same machine
```julia
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  325.000 μs …  2.099 ms  ┊ GC (min … max): 0.00% … 69.74%
 Time  (median):     339.600 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   348.629 μs ± 85.286 μs  ┊ GC (mean ± σ):  1.09% ±  3.81%

     ▁▆█▆▅▄▃▁▁▂▂▁   ▁                                          ▂
  ▇▃▁█████████████████▇▇▇▇▆▆▅▅▅▅▅▆▅▅▅▄▅▅▁▅▄▃▃▃▄▅▄▁▁▃▃▁▄▃▃▄▄▁▁▄ █
  325 μs        Histogram: log(frequency) by time       478 μs <

 Memory estimate: 116.66 KiB, allocs estimate: 5466.
```